Use ARMV8.2 scalar fp16<->fp32 conversion

malfet opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #119909
* __->__ #119895
* #119892

Thanks to discussion with @mikekgfb I've realized that SVE is the
feature availble by default on Apple Silicon, so let use it to speed up
portable but slow bit mashing algorithm implemented as `c10::detail::fp16_ieee_from_fp32_value` by using the following implicit conversion routine:
```cpp
float sve_fp16_to_fp32_value(uint16_t h) {
  union {
     uint16_t h;
     float16_t f16;
  } x = {h};
  return x.f16;
}
```
that according to the https://godbolt.org/z/8s14GvEjo is turned into [`fcvt s0,h0`](https://developer.arm.com/documentation/ddi0596/2021-12/SVE-Instructions/FCVT--Floating-point-convert-precision--predicated--)

As results, very slow and naive [`torch.mm`](https://github.com/pytorch/pytorch/blob/edd9ddf73fae023824c854f23abfe2c15bfcfeee/aten/src/ATen/native/cpu/BlasKernel.cpp#L108) runs 3x faster: 85 msec before to 27 msec (measured by running https://github.com/malfet/llm_experiments/blob/e41341df2d75395277d44e3ae770342fca4bdf18/benchmarks/benchmark_torch_mm.py )

malfet committed:
Use ARMV8.2 scalar fp16<->fp32 conversion

Thanks to discussion with @mikekgfb I've realized that SVE is the
feature availble by default on Apple Silicon, so let use it to speed up
portable but slow bit mashing algorithm implemented as `c10::detail::fp16_ieee_from_fp32_value` by using the following implicit conversion routine:
```cpp
float sve_fp16_to_fp32_value(uint16_t h) {
  union {
     uint16_t h;
     float16_t f16;
  } x = {h};
  return x.f16;
}
```
that according to the https://godbolt.org/z/8s14GvEjo is turned into [`fcvt s0,h0`](https://developer.arm.com/documentation/ddi0596/2021-12/SVE-Instructions/FCVT--Floating-point-convert-precision--predicated--)

[ghstack-poisoned]

mikekgfb reviewed (approved):
How does the ARM_FEATURE_FP16 come into existence?  Can we trust that all compilers that support the feature also have float16_t defined?

Is float16_t __fp16 or _Float16 for ARM?

malfet commented:
> How does the ARM_FEATURE_FP16 come into existence? 

Feature tests macros are defined in https://developer.arm.com/documentation/101028/0012/3--C-language-extensions

> Can we trust that all compilers that support the feature also have float16_t defined?
That I don't know, as float16 becomes a standard only in C++23 (see https://en.cppreference.com/w/cpp/types/floating-point ) 
But again, according to [some ARM docs](https://arm-software.github.io/acle/main/acle.html) 

> The base types are int8_t, uint8_t, int16_t, uint16_t, int32_t, uint32_t, int64_t, uint64_t, float16_t, float32_t, poly8_t, poly16_t, poly64_t, poly128_t and bfloat16_t.

But looks like ISO/IEC TS 18661-3:2015 defines _Float16, so I guess that's the way to go

malfet committed:
Update on "Use ARMV8.2 scalar fp16<->fp32 conversion"


Thanks to discussion with mikekgfb I've realized that SVE is the
feature availble by default on Apple Silicon, so let use it to speed up
portable but slow bit mashing algorithm implemented as `c10::detail::fp16_ieee_from_fp32_value` by using the following implicit conversion routine:
```cpp
float sve_fp16_to_fp32_value(uint16_t h) {
  union {
     uint16_t h;
     float16_t f16;
  } x = {h};
  return x.f16;
}
```
that according to the https://godbolt.org/z/8s14GvEjo is turned into [`fcvt s0,h0`](https://developer.arm.com/documentation/ddi0596/2021-12/SVE-Instructions/FCVT--Floating-point-convert-precision--predicated--)

As results, very slow and naive [`torch.mm`](https://github.com/pytorch/pytorch/blob/edd9ddf73fae023824c854f23abfe2c15bfcfeee/aten/src/ATen/native/cpu/BlasKernel.cpp#L108) runs 3x faster: 85 msec before to 27 msec (measured by running https://github.com/malfet/llm_experiments/blob/e41341df2d75395277d44e3ae770342fca4bdf18/benchmarks/benchmark_torch_mm.py )

[ghstack-poisoned]

malfet commented:
@pytorchbot merge -f "Don't want to wait for periodic builds, as tests are green"

huydhn commented:
@pytorchbot revert -f 'This breaks some RL builds on clang 13, probably due to incompatible clang version header' -f ghfirst

huydhn commented:
@pytorchbot revert -m 'This breaks some RL builds on clang 13, probably due to incompatible clang version header' -f ghfirst

huydhn commented:
@pytorchbot revert -m 'This breaks some RL builds on clang 13, probably due to incompatible clang version header' -c ghfirst

huydhn commented:
It looks like the revert bot couldn't handle the merge on https://github.com/pytorch/pytorch/pull/119909 where it tried to revert a commit from https://github.com/pytorch/pytorch/tree/gh/malfet/67/base and not from base.  As I'm not sure how to remove the top PR from the stack, I guess I could do a fix in merge bot to ignore this case

swolchok commented:
Why do these conversions need to be gated on C10_MOBILE? can't we rely on compiler flags to define an appropriate minimum supported hardware target and then the compiler to respect those flags and not generate illegal instructions?
